### PR TITLE
Extracts a singular docket module

### LINKED ISSUE #86: Pull out docket as class, add helper methods
_This is part of ingest generalization work_

Background:
Scrapers run through a docket of high-level work items (an example of one docket item: search for, scrape, and update any details needed for Person XYZ) for each scrape session.

This is currently baked into the `us_ny` scraper and `queue_control` (soon to be `scraper_control`), but will be a common component of the general scraping infrastructure - so we need to pull out much of that logic, and add it as helper functions in a core `docket` module.

Scope of work:
- Review the design doc's [Life of a Scraper](https://docs.google.com/document/d/11fv-hLwLOFcM5XSE16lYne5hnqFzcQK1vwvTBVVyhn8/edit#heading=h.qz0bdjml50ah) and [Docket](https://docs.google.com/document/d/11fv-hLwLOFcM5XSE16lYne5hnqFzcQK1vwvTBVVyhn8/edit#heading=h.2alazvjvwp9s) sections 
- Review the current docket implementation in `scraper/queue_control.py` and `scraper/us_ny/us_ny_scraper.py`
- Create new `docket.py` module, under `scraper` (soon to be `ingest`) - see design doc [Modules](https://docs.google.com/document/d/11fv-hLwLOFcM5XSE16lYne5hnqFzcQK1vwvTBVVyhn8/edit#heading=h.v6780khuazyt) section
- In `docket` module, provide helper functions for adding, removing items from docket and purging docket for a specific region/scrape type.
- Replace relevant code in `us_ny_scraper` and `queue_control` with calls to the new module.
- Add relevant tests

### PR DESCRIPTION
**Pull request by jessex**
This resolves #86. It extracts docket-specific logic into `docket.py` from `queue_control.py` and `us_ny_scraper.py`. In the process, it also creates a new `sessions.py` that contains session-specific logic that was convenient to extract now (which gets a jump start on #83). 

This includes:
* `docket.py` with all of the docket-specific centralized here into three categories of functionality: population, retrieval, and removal to and from the docket
* `sessions.py`, the beginnings of centralized session-specific logic
* `tracker.py`, a module with logic that spans across `docket` and `sessions` (please suggest a better name`
* A new model, `ScrapeKey`, which contains a `region_code` and a `scrape_type`, representing a conceptual scraper, e.g. (us_ny, background). Adds extra type safety and cleans up code for the various functions in here that take in both of those parameters
* Nearly complete test coverage for the new code, and a significant overall increase
* A handful of bug fixes identified along the way

The NY scraper starts, stops, and resumes successfully for both scrape types locally.

### CODE REVIEW

File: ingest/docket.py [hunk 1]
```diff
@@ -0,0 +1,792 @@
...
+
+# ########################## #
+# Retrieving from the docket #
+# ########################## #
+
+
+def get_new_docket_item(region_code, scrape_type, attempt=0, back_off=5):
+    """Retrieves arbitrary item from docket for the specified scrape type
+
+    Retrieves an arbitrary item still in the docket (whichever docket
```
Comment (ohinds-vera) [inline ingest/docket.py]: Should we add a TODO to optimize this by measuring index updates in practice after we've been running in production for a while?

File: ingest/docket.py [hunk 2]
```diff
@@ -0,0 +1,792 @@
...
+        region_code: (string) Region code (e.g., us_ny)
+        params: List of tuples for other key/value pairs from request
+
+    Returns:
+        N/A
+    """
+    logging.info("Getting target list, %s/%s", region_code, scrape_type)
+
+    if scrape_type == "background":
+        name_list_file = regions.get_name_list_file(region_code)
```
[the remark this answers is not in the document]
Comment (jessex) [reply ingest/docket.py]: Agreed in general. What do you think the scope of that issue would be exactly, though? Removing the requirement for including it in the region manifest? Or providing an alternative path in the extracted docket code? It's hard to change the actual core code path yet without knowing what the alternative path would be.

File: ingest/docket.py [hunk 3]
```diff
@@ -0,0 +1,792 @@
...
+        phase = 0
+        load_snapshot_target_list(region_code, phase)
+
+
+def load_background_target_list(region_code, name_file, query_name, load,
+                                start_line=1):
+    """Load background scrape docket items, from name file.
+
+    Iterates over a CSV of common names, loading a docket item for the scraper
+    to search for each one. We use pandas because it can selectively load lines
```
[the remark this answers is not in the document]
Comment (jessex) [reply ingest/docket.py]: I didn't review this docstring closely when I copied it from elsewhere. I'll do some archaeological work and see if it was ever true or just a mistake. Either way, the docs will be made accurate here.

File: ingest/docket.py [hunk 4]
```diff
@@ -0,0 +1,792 @@
...
+            batch_size, start_cursor=cursor)
+        print query_results, next_cursor, more, phase
+
+    else:
+        logging.info("Finished loading snapshot target list to docket.")
+        return
+
+    (relevant_records, inmate_keys_ids) = process_query_response(
+        query_type, query_results)
+    print relevant_records
```
[the remark this answers is not in the document]
Comment (jessex) [reply ingest/docket.py]: Yes they are, thanks! 👀

File: ingest/docket.py (hunk 1 above)
Comment (jessex) [reply ingest/docket.py]: Good question. I think there are a number of potential optimizations that we could evaluate going forward that are not captured yet. I'll create a new `optimization` label and file some issues tomorrow, including this one.

**Comment by jessex** [comment]
@ohinds-vera: I just pushed up the two minor updates requested. I still need to file the issues you called out and will do so this week, but we should be ready to merge here with your approval. Do note that once this is merged, I would predict non-trivial merge conflicts for you, depending on how much code has moved around on your branch. Would be happy to help you sort through it all, but would like to get this merged so we can proceed.

**Pull request merged**